client/src/controllers/FormsetController.ts

```diff
@@ -0,0 +1,283 @@
+  declare readonly templateTarget: HTMLTemplateElement;
+  /** Hidden input to track the total forms (including deleted) for POST request and initial reading. */
+  declare readonly totalFormsInputTarget: HTMLInputElement;
+  /** Set to the value to the management field MIN_NUM_FORMS. */
+  declare minValue: number;
+  /** Set to the value to the management field MAX_NUM_FORMS. */
+  declare maxValue: number;
+  /** Value tracking for the total amount of forms either active or deleted. */
+  declare totalValue: number;
+
+  initialize() {
+    this.totalValue = parseInt(this.totalFormsInputTarget.value, 10);
+    this.minValue = parseInt(this.minFormsInputTarget.value, 10);
+    this.maxValue = parseInt(this.maxFormsInputTarget.value, 10);
+  }
+
+  connect() {
+    this.dispatch('ready', {
+      cancelable: false,
+      detail: {
+        minValue: this.minValue,
+        maxValue: this.maxValue,
+        totalValue: this.totalValue,
+      },
+    });
+  }
+
+  /**
+   * Add a new child form from the template content.
+   */
+  add() {
+    if (this.childTargets.length >= this.maxValue) return;
+
+    if (
+      this.dispatch('adding', {
+        cancelable: true,
+        detail: { formIndex: this.totalValue },
+      }).defaultPrevented
+    ) {
+      return;
+    }
+
+    this.formsTarget.append(this.newChild);
+  }
+
+  /**
+   * Find the event's target's closest child target and remove it by
+   * removing the 'child' target and adding a 'child-removed' target.
+   */
+  delete(event: Event) {
+    const target = this.childTargets.find((child) =>
+      child.contains(event.target as Node),
+    );
+
+    if (!target) {
+      throw new Error(
+        `Could not find child form target for event target: ${event.target}.`,
+      );
+    }
+
+    if (this.childTargets.length <= this.minValue) return;
+
+    if (
+      this.dispatch('removing', { target, cancelable: true }).defaultPrevented
+    ) {
+      return;
+    }
+
+    const targetAttrName = `data-${this.identifier}-target`;
+
+    target.setAttribute(
+      targetAttrName,
+      (target.getAttribute(targetAttrName)?.split(' ') ?? [])
+        .filter((name) => name !== 'child')
+        .concat(['deleted'])
+        .join(' '),
+    );
+  }
+
+  /**
+   * When a new child is added, update the total count and dispatch an added event.
+   */
+  childTargetConnected(target: HTMLElement) {
+    const totalFormsCount =
+      this.childTargets.length + this.deletedTargets.length;
+    if (totalFormsCount === this.totalValue) return;
+    this.totalValue = totalFormsCount;
+    this.dispatch('added', {
+      target,
+      cancelable: false,
+      detail: { formIndex: totalFormsCount - 1 },
+    });
+  }
+
+  /**
+   * When removed, add the class and update the total count.
+   * Also update the DELETE input for this form.
+   */
+  childTargetDisconnected(target: HTMLElement) {
+    const deletedClasses = this.deletedClasses;
+
```

Comment:
We can probably check if the element is already hidden and bypass this logic. Might be a better and simpler approach to ensure no events get dispatched on POST error scenarios when the deleted items may already be in the response.